### PR TITLE
Add extension for Do Not Disturb Mode (and underlay User Preference extension)

### PR DESCRIPTION
**Pull request by waynezhang**
Add a new extension to control Do Not Disturb Mode

```lua
-- Get current value
local status = hs.donotdisturb.status()

-- Turn on  Do Not Disturb Mode
hs.donotdisturb.on()

-- Turn off  Do Not Disturb Mode
hs.donotdisturb.off()
```

And because it's depended on user preference, I also isolated it as a standalone extension

```lua
-- Set preference
hs.userpreferences.set("key", value, "application id")

-- Get preference
hs.userpreferences.get("key", "application id")

-- Synchronize preferences
hs.userpreferences.sync("application id")
```

### CODE REVIEW

**Comment by latenitefilms** [comment]
Amazing work @waynezhang!

My only question is, how does `hs.userpreference` fit in with @cmsj's `hs.plist` and @asmagill's WIP `hs.preferences`?

See issues: 

- #1498 - `hs.preferences` - Property List Editor with `NSKeyedArchiver` support 
- #1441 - `CFPreferences`
- #1432 - Extension for Reading/Writing External Property List Files 

Thoughts @asmagill & @cmsj ?

**Comment by cmsj** [comment]
(will report some thoughts later, just wanted to quickly drop in to say ignore the sticker CI comments, it doesn't know what it's talking about)

**Comment by latenitefilms** [comment]
@cmsj - I've just made a tweak in #1707 which I think will solve the @stickler-ci errors.

File: extensions/userpreference/internal.m [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    NSString *key = [skin toNSObjectAtIndex:1];
+    NSString *applicationId = [skin toNSObjectAtIndex:2];
+
+    CFPropertyListRef ref = CFPreferencesCopyValue((__bridge CFStringRef)key,
+                                                   (__bridge CFStringRef)applicationId,
+                                                   kCFPreferencesCurrentUser,
+                                                   kCFPreferencesCurrentHost);
+    if (ref != nil) {
+        [skin pushNSObject:(__bridge NSObject *)ref];
```
Comment (cmsj) [inline extensions/userpreference/internal.m]: Small thing, but we tend to use `__bridge_transfer` rather than `__bridge` and `CFAutorelease()`.

**Comment by dbalatero** [comment]
@waynezhang Would love to see this merged so I can control DND from the keyboard. Anything I can do to help?

**Comment by waynezhang** [comment]
@dbalatero Sorry for late reply. I'm gonna fix this recently when I have time. The tasks left will be:
- Rename some variable according to @cmsj's comment
- To see if user preference related stuff can be replaced by the PR mentioned by @latenitefilms 
I'd be very appreciated if you can lend me some help :)

**Comment by megalithic** [comment]
@waynezhang any updates/thoughts around resolving these items so @cmsj can merge?

**Comment by latenitefilms** [comment]
I'd love to get @asmagill's thoughts in regards to `hs.userpreferences`, and how that fits in with the other Preferences/Settings extensions?

**Comment by sjkyspa** [comment]
when will this merged?

**Comment by latenitefilms** [comment]
@sjkyspa - The big question here is how to deal with the preferences handling.

What do you think @cmsj & @asmagill ?

Related:

#1498 - hs.preferences - Property List Editor with NSKeyedArchiver support
#1441 - CFPreferences

**Comment by latenitefilms** [comment]
@cmsj & @asmagill - In the interest of merging in the code to solve the original issue in this pull request, I wonder if it's worth just removing `hs.userpreference` for now, and hardcoding in the preferences stuff directly in `hs.donotdisturb` so this can be merged, and we can work out what to do with plists in another issue/pull request?

**Comment by asmagill** [comment]
Sorry I didn't look at this earlier... @latenitefilms since you are the only one I know has used hs._asm.cfpreferences much beyond my initial tests, what are your thoughts on it? What does it need/lack before being submitted for inclusion?

It looks like userpreferences here is a subset, so we could create a wrapper (or even rename mine) if the name seems to make more sense... and since I do note in cfpreferences's limitations that it can't access the "AnyUser" key for writing, renaming mine might actually be worthwhile... the label "cfpreferences" is just a nod to the underlying foundation library being used and wouldn't make sense to someone not somewhat familiar with macOS programming...

**Comment by latenitefilms** [comment]
Currently I'm using a mixture of `hs._asm.cfpreferences`, `hs.plist` and our own Lua-based `cp.plist` (which internally uses `plutil` and `openssl base64`) to achieve everything we need. I'll need to go back through our code to work out the pro's and con's of each extension.

From memory, the reason we're using `cp.plist` in a lot of cases is that we're also using `cp.plist.archiver`, which supports 'defrosting' a table which is made up from an `NSKeyArchiver` record. We can't use `hs.pasteboard.readArchiverDataForUTI()` because we're dealing with classes that Hammerspoon doesn't support.

There were also cases where `hs._asm.cfpreferences` was giving different results than `cp.plist`. I can't remember the specifics off the top of my head.

Ideally, I'd love to drop `cp.plist` and use faster Objective-C implementations.

I think renaming `hs._asm.cfpreferences` to `hs.userpreferences` makes sense?

Note to self - these plugins use `cp.plist`:

- `plugins.finalcutpro.notifications.manager`
- `plugins.finalcutpro.pasteboard.manager`
- `plugins.finalcutpro.timeline.commandsetactions`
- `cp.apple.finalcutpro.export.destinations`
- `cp.apple.finalcutpro.plugins`
- `cp.disk`

These plugins use `cp.plist.archiver`:

- `cp.app.menu`
- `cp.apple.finalcutpro.export.destinations`
- `cp.apple.finalcutpro.plugins`
- `plugins.finalcutpro.pasteboard.manager`

Also, note to self, that `cp.apple.finalcutpro.export.destinations` says:

```lua
----------------------------------------------------------------------------------
-- TODO: This code currently doesn't work with hs.plist, which is why we're still
--       using cp.plist. Will re-investigate once we have hs.plist.readString()
----------------------------------------------------------------------------------
```

Will do some experimenting over the coming days to work out what use-cases `hs._asm.cfpreferences` is currently failing for us.

**Comment by latenitefilms** [comment]
@asmagill - Re-looking at this issue, the `hs.userpreference` extension added in this pull request can be easily replaced with `hs._asm.cfpreferences`.

For example:

```lua
> cfpreferences = require("hs._asm.cfpreferences")


> cfpreferences.getValue("doNotDisturb", "com.apple.notificationcenterui")
false

> cfpreferences.getValue("doNotDisturb", "com.apple.notificationcenterui")
true
```

Given that, I'd suggest we try and merge `hs._asm.cfpreferences` into Hammerspoon at some point when you get time (obviously there's a fair amount of work to do, as `hs._asm.cfpreferences` is currently undocumented, etc. - I'm happy to help where I can!), then once that's done this pull request just requires a few minor tweaks before it can be merged in. This could either be done by @waynezhang, or I'm happy to help make the necessary changes, once `hs._asm.cfpreferences` is merged in.

As discussed above, I think renaming `hs._asm.cfpreferences` to `hs.userpreferences` also makes sense.

**Comment by waynezhang** [comment]
Sorry for leaving this issue untouched for so long a time.
I'll make some fix soon.

@latenitefilms Thanks for the advice. I'm happy to do the migration to `_asm.cfpreferences` after it's merged to Hammerspoon.

File: extensions/userpreference/internal.m (hunk 1 above)
Comment (waynezhang) [reply extensions/userpreference/internal.m]: Fixed at https://github.com/Hammerspoon/hammerspoon/pull/1706/commits/f833676a6e1b86cb895ea15c8c07aac4da8d7b35.

**Comment by waynezhang** [comment]
I've renamed it to `hs.donotdisturb` (and `hs.userpreference` to `hs.userpreferences`). PTAL.

**Comment by evantravers** [comment]
Nothing to add here but excited to see this 👍

**Comment by asmagill** [comment]
I want to leave this for 0.9.80 -- I plan to work on the merging of cfpreferences, preferences, settings, etc. after getting 0.9.79 out the door and this would be a perfect fit at that time.

**Comment by Nezteb** [comment]
I'd love this feature; for the time being I'm using this method: https://spinscale.de/posts/2020-10-30-using-hammerspoon-do-not-disturb-screen-sharing-zoom.html

**Comment by evantravers** [comment]
For anyone else thinking about this in the meantime: if you are on Big Sur you can trigger a shortcut using `hs.shortcuts.run` to enable Focus Modes or DND. You just have to create the corresponding Shortcuts.

https://github.com/evantravers/hammerspoon-config/blob/219c0fa903ca2c0bafbad2a6564f1c4feb7427f6/spaces/deep.lua#L14

For now… this works really well for me, and I'm ok.

**Pull request closed without merging**